src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -223,65 +243,84 @@ protected ImplementationInfo computeValue(Class<?> intfc) {
             MethodType[] types = new MethodType[methods.size()];
             for (int i = 0; i < methods.size(); i++) {
                 Method sm = methods.get(i);
-                MethodType mt = MethodType.methodType(sm.getReturnType(), sm.getParameterTypes());
+                MethodType mt = methodType(sm.getReturnType(), sm.getParameterTypes());
                 types[i] = mt;
                 var thrown = sm.getExceptionTypes();
-                var fieldName = "m" + i;
                 if (thrown.length == 0) {
-                    infos.add(new LocalMethodInfo(fieldName, mt, desc(mt), DEFAULT_RETHROWNS));
+                    infos.add(new LocalMethodInfo(desc(mt), DEFAULT_RETHROWS));
                 } else {
-                    infos.add(new LocalMethodInfo(fieldName, mt, desc(mt), Stream.concat(Stream.of(RuntimeException.class, Error.class),
-                            Arrays.stream(thrown)).map(MethodHandleProxies::desc).distinct().toList()));
+                    infos.add(new LocalMethodInfo(desc(mt), Stream.concat(DEFAULT_RETHROWS.stream(),
+                            Arrays.stream(thrown).map(MethodHandleProxies::desc)).distinct().toList()));
                 }
             }
 
-            var bytes = spin(desc(intfc), methods.get(0).getName(), infos);
-            Class<?> resultClass;
-            MethodHandle ctor, getType, getTarget;
+            // default class hierarchy resolver accesses system resources
+            @SuppressWarnings("removal")
+            var sm = System.getSecurityManager();
+            @SuppressWarnings("removal")
+            byte[] template = sm != null ? AccessController.doPrivileged(new PrivilegedAction<byte[]>() {
+                @Override
+                public byte[] run() {
+                    return createTemplate(desc(intfc), methods.get(0).getName(), infos);
+                }
+            }) : createTemplate(desc(intfc), methods.get(0).getName(), infos);
+
+            return new InterfaceInfo(types, new Lookup(intfc), template);
+        }
+    };
 
+    private static final ClassValue<WrapperInfo> WRAPPER_INFOS = new ClassValue<>() {
+        @Override
+        protected WrapperInfo computeValue(Class<?> type) {
+            // WrapperInstance is in non-exported package
+            @SuppressWarnings("removal")
+            var sm = System.getSecurityManager();
+            @SuppressWarnings("removal")
+            WrapperInstance anno = sm != null ? AccessController.doPrivileged(new PrivilegedAction<>() {
+                @Override
+                public WrapperInstance run() {
+                    return type.getDeclaredAnnotation(WrapperInstance.class);
+                }
+            }) : type.getDeclaredAnnotation(WrapperInstance.class);
```

Comment:
I'm not sure how the example shows that this is a security vulnerability? It still works fine without the call to `isWrapperInstance` (even if you switch to using jdk.internal.misc.Unsafe.class, although that also requires `--add-exports` when compiling)

Reply:
Guess I will just nuke the annotation and check for its implemented clause instead.

Reply:
I don't think that's good enough. If an instance implements an interface that just happens to also have an entry in the INTERFACE_INFOS cache, then e.g. `MethodHandleProxies::wrapperInstanceTarget` could be used to retrieve the class data of a class that wasn't generated using MethodHandleProxies.

I think some explicit marker is needed for class generated by MethodHandleProxies.

Switching back to the `WrapperInstance` interface, and then defining the class using `IMPL_LOOKUP`, as opposed to defining the class in a dynamic module, could be an option (since we determined that the wrapper instance can not be used to escalate privilege. I'm just not sure if that will work for target interfaces in non-exported packages. Proxy seems to be able to do that too though.

Reply:
An other possibility to to add some key into the ClassData that can't be imitated - such as the `InterfaceInfo` itself.  
An `instanceof` check on the ClassData at a fixed position could then determine if is a MHP.

Reply:
It looks like Proxy forcibly adds reads and exports edges from the module of the target interface to the module that defines the proxy: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/reflect/Proxy.java#L855-L868 I think we can do the same here

Reply:
I would support the putting InterfaceInfo in class data approach. With IMPL_LOOKUP, we currently define the hidden class in the same module as the interface. It's currently hard to define a hidden class without an anchor host class for a new module.